pkg/config/app_configuration.go

```diff
@@ -13,6 +13,13 @@ type ApplicationConfig struct {
 	// Duration. example: "30s"
 	ActorScanInterval string `json:"actorScanInterval"`
 	// Duration. example: "30s"
-	DrainOngoingCallTimeout string `json:"drainOngoingCallTimeout"`
-	DrainRebalancedActors   bool   `json:"drainRebalancedActors"`
+	DrainOngoingCallTimeout string           `json:"drainOngoingCallTimeout"`
+	DrainRebalancedActors   bool             `json:"drainRebalancedActors"`
+	Reentrancy              ReentrancyConfig `json:"reentrancy,omitempty"`
+}
+
+type ReentrancyConfig struct {
+	Enabled bool     `json:"enabled"`
+	Methods []string `json:"methods,omitempty"`
```

Comment:
Do we really need this to be per method? Can't it be all methods?

Reply:
This was a topic that came up in the design discussion. We decided it'd be nice to be able to specify, but the default is all methods are reentrant unless otherwise specified. Essentially, an empty (or missing) array means everything is ready, if the array is specified, only those methods are allowed.

If you feel that this is complexity we don't need right now, I can understand that and am open to adding it as an incremental improvement down the road, what are your thoughts?

Reply:
I am more worried about how we support this. Can we remove this for now? I like the idea of adding it as an incremental improvement later.

Reply:
I'd also like to get @rynowak's opinion. I'm on the fence about it implemented now vs. later. I personally probably won't use this feature, but I can think of a couple of methods I wouldn't want reentrency. If I'm understanding this correctly, the allowed methods would need  to be known at startup; so it's unlikely PHP could make good use of this feature (there's no way to scan the code for attributes, so it would need to be a handwritten list on the PHP side).

It might be simpler for SDK's to expose whether a call is a reentrant call on a context and let the application author decide whether to service the method or not.

Reply:
@rynowak ?

Reply:
> It might be simpler for SDK's to expose whether a call is a reentrant call on a context and let the application author decide whether to service the method or not.

This seems like the most useful and flexible semantic, and also one of the easiest to reason about from the POV of the code author 

"I am about to start some work that will call back into this actor" is a far more connected to the code than "this method can be called reentrantly"

Reply:
Weird, I commented on this - it showed up as a double-post and then when I deleted one of them it deleted both 😆 

From the POV of something like C# it would be easy for us to detect w/ an attribute. For instance we'll likely have an attribute on the class level to enable reentrancy. If we go this way it's no trouble to have a method-level attribute. 

If I needed fine-grained control, the thing that I'd probably most want as a user would be the ability (per-call) to say that my outgoing call *may* trigger reentrant processing. That's much more fine-grained and flexible than having an allow-list of methods. I decide which *states* are valid for reentrant processing rather than which entry-points are valid for reentrant processing.

> so it would need to be a handwritten list on the PHP 

How does actor registration work in PHP - would you make this an additional parameter/data when registering the actor type?

Reply:
Based on the conversation here, I think it'd be better to implement methods later, especially given the level of control that Ryan is mentioning. It seems like we could have a larger discussion around it and what our intent with it is.